### ISSUE TITLE
unable to load model from https://docs.opencv.org/3.3.1/de/d37/tutorial_dnn_halide.html

### ISSUE THREAD

**Issue by guoyejun** [opencv/opencv#14829]
##### System information (version)

- OpenCV => tag 4.1.0 
- Operating System / Platform => ubuntu 16.04 x64
- Compiler => gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.11) 

##### Detailed description

i have installed llvm/halide, and build opencv with halide support (with debug version).
i use the source code exactly the same as https://docs.opencv.org/3.3.1/de/d37/tutorial_dnn_halide.html

and download the files from https://raw.githubusercontent.com/DeepScale/SqueezeNet/master/SqueezeNet_v1.1/train_val.prototxt and https://github.com/DeepScale/SqueezeNet/raw/master/SqueezeNet_v1.1/squeezenet_v1.1.caffemodel

##### Steps to reproduce

just to run the application compiled from the above source code, it crashes. And so gdb it to get the bt:
...
#4  0x00007ffff5a66701 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff5a66919 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007ffff5f6af51 in cv::error (exc=...) at /work/media/opencv_work/opencv/modules/core/src/system.cpp:1032
#7  0x00007ffff5f6b01e in cv::error (_code=-5, _err="Duplicate blobs produced by multiple sources", _func=0x7ffff7a35c10 <cv::dnn::dnn
4_v20190122::(anonymous namespace)::CaffeImporter::addOutput(opencv_caffe::LayerParameter const&, int, int)::__func__> "addOutput", _file=0x7ffff7a35470 "/work/media/opencv_work/opencv/modules/dnn/src/caffe/caffe_importer.cpp", _line=427) at /work/media/opencv_work/opencv/modules/core/src/system.cpp:1044
#8  0x00007ffff7786f9f in cv::dnn::dnn4_v20190122::(anonymous namespace)::CaffeImporter::addOutput (this=0x7fffffffdd10, layer=..., layerId=2, outNum=0) at /work/media/opencv_work/opencv/modules/dnn/src/caffe/caffe_importer.cpp:427
#9  0x00007ffff7786ae2 in cv::dnn::dnn4_v20190122::(anonymous namespace)::CaffeImporter::populateNet (this=0x7fffffffdd10, dstNet=...)
 at /work/media/opencv_work/opencv/modules/dnn/src/caffe/caffe_importer.cpp:402
#10 0x00007ffff77872e9 in cv::dnn::dnn4_v20190122::readNetFromCaffe (prototxt="train_val.prototxt", caffeModel="squeezenet_v1.1.caffemodel") at /work/media/opencv_work/opencv/modules/dnn/src/caffe/caffe_importer.cpp:458
#11 0x0000000000404c22 in main (argc=1, argv=0x7fffffffe1f8) at /work/media/opencv_work/halide_try/classprob.cpp:43

**Comment by alalek** [comment]
At first we need to find reason of the failure.
Error message is related to model importer, so probably Halide can't affect that.
Try to reduce dependencies and exclude Halide from the build.

Did you try to run OpenCV DNN tests?
This is separate test for [SqueezeNet 1.1](https://github.com/opencv/opencv/blob/3.4.6/modules/dnn/test/test_caffe_importer.cpp#L306-L307).

**Comment by guoyejun** [comment]
thanks, yes, it should has nothing to do with halide, i just reported this issue to see if anyone happens to know the reason.

after sometime debug, it turns out the issue is that the importer is unable to handle the following proto, it doesn't analyze it is for TRAIN or TEST, so the importer reports error that there are two layers with the same output.
layer {
  name: "data"
  type: "Data"
  top: "data"
  top: "label"
  include {
    phase: TRAIN
  }
  transform_param {
    crop_size: 227
    mean_value: 104
    mean_value: 117
    mean_value: 123
  }
  data_param {
    source: "examples/imagenet/ilsvrc12_train_lmdb"
    batch_size: 32
    backend: LMDB
  }
}
layer {
  name: "data"
  type: "Data"
  top: "data"
  top: "label"
  include {
    phase: TEST
  }
  transform_param {
    crop_size: 227
    mean_value: 104
    mean_value: 117
    mean_value: 123
  }
  data_param {
    source: "examples/imagenet/ilsvrc12_val_lmdb"
    batch_size: 25 #not *iter_size
    backend: LMDB
  }
}

the simple method is to change https://docs.opencv.org/3.3.1/de/d37/tutorial_dnn_halide.html to use https://raw.githubusercontent.com/opencv/opencv_extra/master/testdata/dnn/squeezenet_v1.1.prototxt instead of https://raw.githubusercontent.com/DeepScale/SqueezeNet/master/SqueezeNet_v1.1/train_val.prototxt

**Comment by alalek** [comment]
OpenCV can't process training configurations like `SqueezeNet_v1.1/train_val.prototxt`.

In general, you should pass "deploy" (inference) configuration instead: [SqueezeNet_v1.1/deploy.prototxt](https://github.com/DeepScale/SqueezeNet/blob/master/SqueezeNet_v1.1/deploy.prototxt) (but it requires some fixes too).

Documentation sample with "train_val" has been removed from latest [`3.4` branch](https://docs.opencv.org/3.4/de/d37/tutorial_dnn_halide.html).

Probably we should add `phase: TRAIN` detection into OpenCV and emitting an error about misused configuration.

**Comment by guoyejun** [comment]
thanks,  

but, just curious why we remove the example from the document? We can just choose another example with the good model file. 

I tried this example again with https://github.com/opencv/opencv_extra/blob/master/testdata/dnn/squeezenet_v1.1.prototxt on OpenCV 4.1.0, the network model is successfully loaded, but there is a crash within the Halide module. (I'm using halide latest code with LLVM 8.0).
`(gdb) bt
#0  0x00007ffff7fdc3ea in pool1 ()
#1  0x00007ffff7fdca73 in pool1_argv ()
#2  0x00007fffed4a8cfe in Halide::Pipeline::call_jit_code(Halide::Target const&, Halide::Pipeline::JITCallArgs const&) () from /work/me
dia/halide_work/Halide/build/lib/libHalide.so
#3  0x00007fffed4a9037 in Halide::Pipeline::realize(Halide::Pipeline::RealizationArg, Halide::Target const&, Halide::ParamMap const&) (
) from /work/media/halide_work/Halide/build/lib/libHalide.so
#4  0x00007fffed2947f1 in Halide::Func::realize(Halide::Pipeline::RealizationArg, Halide::Target const&, Halide::ParamMap const&) () fr
om /work/media/halide_work/Halide/build/lib/libHalide.so
#5  0x00007ffff777d95e in cv::dnn::forwardHalide (outputs=std::vector of length 1, capacity 1 = {...}, node=...) at /work/media/opencv_
work/opencv/modules/dnn/src/op_halide.cpp:216
#6  0x00007ffff7737ed0 in cv::dnn::dnn4_v20190122::Net::Impl::forwardLayer (this=0x66c070, ld=...) at /work/media/opencv_work/opencv/mo
dules/dnn/src/dnn.cpp:2499
#7  0x00007ffff77388b9 in cv::dnn::dnn4_v20190122::Net::Impl::forwardToLayer (this=0x66c070, ld=..., clearFlags=true) at /work/media/op
encv_work/opencv/modules/dnn/src/dnn.cpp:2555
#8  0x00007ffff7739d82 in cv::dnn::dnn4_v20190122::Net::forward (this=0x7fffffffdee0, outputName="prob") at /work/media/opencv_work/ope
ncv/modules/dnn/src/dnn.cpp:2779
#9  0x0000000000404fdb in main (argc=1, argv=0x7fffffffe1f8) at /work/media/opencv_work/halide_try/classprob.cpp:68`

anyway, the tests passed.
`$ ./opencv_test_dnn --gtest_filter=Layer_Test_Halide*
...
[----------] Global test environment tear-down
[==========] 618 tests from 14 test cases ran. (1768020 ms total)
[  PASSED  ] 618 tests.`

what can i try? My purpose is just to write a simple but complete OpenCV application that uses the halide backend. thanks.

**Comment by dkurt** [comment]
@guoyejun, Please try to run networks tests as well (--gtest_filter=DNNTestNetwork*)

**Comment by guoyejun** [comment]
'''
build/bin$ ./opencv_test_dnn --gtest_filter=DNNTestNetwork*
[ INFO:0] Initialize OpenCL runtime...
CTEST_FULL_OUTPUT
OpenCV version: 4.1.0
OpenCV VCS version: 4.1.0
Build type: Debug
Compiler: /usr/bin/c++  (ver 5.4.0)
Parallel framework: pthreads
CPU features: SSE SSE2 SSE3 *SSE4.1 *SSE4.2 *FP16 *AVX *AVX2 *AVX512-SKX?
OpenCL is disabled
Note: Google Test filter = DNNTestNetwork*
[==========] Running 20 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 20 tests from DNNTestNetwork
[ RUN      ] DNNTestNetwork.AlexNet/0, where GetParam() = HALIDE/CPU
[     SKIP ] OpenCV tests: Can't find data file: dnn/bvlc_alexnet.caffemodel
[       OK ] DNNTestNetwork.AlexNet/0 (1 ms)
[ RUN      ] DNNTestNetwork.ResNet_50/0, where GetParam() = HALIDE/CPU
[     SKIP ] OpenCV tests: Can't find data file: dnn/ResNet-50-model.caffemodel
[       OK ] DNNTestNetwork.ResNet_50/0 (1 ms)
[ RUN      ] DNNTestNetwork.SqueezeNet_v1_1/0, where GetParam() = HALIDE/CPU
[     SKIP ] OpenCV tests: Can't find data file: dnn/squeezenet_v1.1.caffemodel
[       OK ] DNNTestNetwork.SqueezeNet_v1_1/0 (1 ms)
[ RUN      ] DNNTestNetwork.GoogLeNet/0, where GetParam() = HALIDE/CPU
[     SKIP ] OpenCV tests: Can't find data file: dnn/bvlc_googlenet.caffemodel
[       OK ] DNNTestNetwork.GoogLeNet/0 (2 ms)
[ RUN      ] DNNTestNetwork.Inception_5h/0, where GetParam() = HALIDE/CPU
[     SKIP ] OpenCV tests: Can't find data file: dnn/tensorflow_inception_graph.pb
[       OK ] DNNTestNetwork.Inception_5h/0 (1 ms)
[ RUN      ] DNNTestNetwork.ENet/0, where GetParam() = HALIDE/CPU
[     SKIP ] OpenCV tests: Can't find data file: dnn/Enet-model-best.net
[       OK ] DNNTestNetwork.ENet/0 (6 ms)
[ RUN      ] DNNTestNetwork.MobileNet_SSD_Caffe/0, where GetParam() = HALIDE/CPU
[     SKIP ] 
[       OK ] DNNTestNetwork.MobileNet_SSD_Caffe/0 (0 ms)
[ RUN      ] DNNTestNetwork.MobileNet_SSD_Caffe_Different_Width_Height/0, where GetParam() = HALIDE/CPU
[     SKIP ] 
[       OK ] DNNTestNetwork.MobileNet_SSD_Caffe_Different_Width_Height/0 (0 ms)
[ RUN      ] DNNTestNetwork.MobileNet_SSD_v1_TensorFlow/0, where GetParam() = HALIDE/CPU
[     SKIP ] 
[       OK ] DNNTestNetwork.MobileNet_SSD_v1_TensorFlow/0 (0 ms)
[ RUN      ] DNNTestNetwork.MobileNet_SSD_v1_TensorFlow_Different_Width_Height/0, where GetParam() = HALIDE/CPU
[     SKIP ] 
[       OK ] DNNTestNetwork.MobileNet_SSD_v1_TensorFlow_Different_Width_Height/0 (0 ms)
[ RUN      ] DNNTestNetwork.MobileNet_SSD_v2_TensorFlow/0, where GetParam() = HALIDE/CPU
[     SKIP ] 
[       OK ] DNNTestNetwork.MobileNet_SSD_v2_TensorFlow/0 (0 ms)
[ RUN      ] DNNTestNetwork.SSD_VGG16/0, where GetParam() = HALIDE/CPU
[     SKIP ] 
[       OK ] DNNTestNetwork.SSD_VGG16/0 (0 ms)
[ RUN      ] DNNTestNetwork.OpenPose_pose_coco/0, where GetParam() = HALIDE/CPU
[     SKIP ] 
[       OK ] DNNTestNetwork.OpenPose_pose_coco/0 (0 ms)
[ RUN      ] DNNTestNetwork.OpenPose_pose_mpi/0, where GetParam() = HALIDE/CPU
[     SKIP ] 
[       OK ] DNNTestNetwork.OpenPose_pose_mpi/0 (0 ms)
[ RUN      ] DNNTestNetwork.OpenPose_pose_mpi_faster_4_stages/0, where GetParam() = HALIDE/CPU
[     SKIP ] 
[       OK ] DNNTestNetwork.OpenPose_pose_mpi_faster_4_stages/0 (0 ms)
[ RUN      ] DNNTestNetwork.OpenFace/0, where GetParam() = HALIDE/CPU
[     SKIP ] 
[       OK ] DNNTestNetwork.OpenFace/0 (0 ms)
[ RUN      ] DNNTestNetwork.opencv_face_detector/0, where GetParam() = HALIDE/CPU
[     SKIP ] 
[       OK ] DNNTestNetwork.opencv_face_detector/0 (0 ms)
[ RUN      ] DNNTestNetwork.Inception_v2_SSD_TensorFlow/0, where GetParam() = HALIDE/CPU
[     SKIP ] 
[       OK ] DNNTestNetwork.Inception_v2_SSD_TensorFlow/0 (0 ms)
[ RUN      ] DNNTestNetwork.DenseNet_121/0, where GetParam() = HALIDE/CPU
[     SKIP ] 
[       OK ] DNNTestNetwork.DenseNet_121/0 (0 ms)
[ RUN      ] DNNTestNetwork.FastNeuralStyle_eccv16/0, where GetParam() = HALIDE/CPU
[     SKIP ] 
[       OK ] DNNTestNetwork.FastNeuralStyle_eccv16/0 (0 ms)
[----------] 20 tests from DNNTestNetwork (12 ms total)

[----------] Global test environment tear-down
[==========] 20 tests from 1 test case ran. (13 ms total)
[  PASSED  ] 20 tests.
'''

I have done git clone opencv_extra and export OPENCV_TEST_DATA_PATH=/work/media/opencv_work/opencv_extra/testdata/

**Comment by alalek** [comment]
Actually all these tests are skipped due missing models.
There is downloading script in opencv_extra: [download_models.py](https://github.com/opencv/opencv_extra/blob/4.1.0/testdata/dnn/download_models.py).
All models requires about 5Gb (you may edit this file to load subset only - AlexNet is a good start).

You can run this script from your `opencv_extra/testdata/dnn/` location or use different one:
- with 'dnn' nested folder: `<models_path>/dnn`
- configure `OPENCV_DNN_TEST_DATA_PATH=<models_path>` environment variable
